### PR TITLE
fix(CL): error blow up in CalcAmount0Delta causes OverChargeSwapOutGivenInError when swapping zero for one

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #6351 

## What is the purpose of the change

See #6351 for detailed context.

Our current order of math operations in `CalcAmount0Delta` was causing error blow-up. This was noticed pre-CL launch but was deemed as acceptable due to occurring infrequently.

By lowering our min spot price to 10^-30, the frequency of this error occurring was increased to an unacceptable degree.

The core of the issue is that `a / (b * c)` amplifies the error when b and c are both small. Division by extremely small numbers has such an effect.

By changing the order of operations to `a / b / c `, we reduce the error amplification and resolve the problem encountered.

The test added on this branch reproduces the edge case (expected results are estimated in Python).

Made changes to the new test helper in `osmoassert` as a drive-by change

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/math/math.go [hunk 1]
```diff
@@ -72,8 +72,7 @@ func CalcAmount0Delta(liq, sqrtPriceA, sqrtPriceB osmomath.BigDec, roundUp bool)
 		// - calculating amountIn during swap
 		// - adding liquidity (request user to provide more tokens in in favor of the pool)
 		// The denominator is truncated to get a higher final amount.
-		denom := sqrtPriceA.MulTruncate(sqrtPriceB)
-		return liq.Mul(diff).QuoMut(denom).Ceil()
+		return liq.MulRoundUp(diff).QuoTruncate(sqrtPriceA).QuoTruncate(sqrtPriceB).Ceil()
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/math/math.go:75]: What was the reasoning behind having the divisions be QuoTruncate if we're trying to round up? My understanding is this truncates at precision and thus rounds those operations down

**Review by AlpinYukseloglu — APPROVED** [review]
LGTM, nice job finding a clean fix for this. I left a non-blocking comment about rounding direction (still approving since rounding direction at precision doesn't matter here as long as we Ceil at the end)

File: x/concentrated-liquidity/math/math.go (hunk 1 above)
Comment (p0mvn) [reply x/concentrated-liquidity/math/math.go]: Thanks for catching this! I've been staring at this way too long and inverted the direction in my head. Fixed here:
[f149a90](https://github.com/osmosis-labs/osmosis/pull/6352/commits/f149a903967ffd2094744d86f2b7eb96cf7f84ae)

**Comment by p0mvn** [comment]
@AlpinYukseloglu  Thanks for the review, made some extra updates to apply the same order of operations when rounding down:
[f8fb807](https://github.com/osmosis-labs/osmosis/pull/6352/commits/f8fb807e9269fad20d611a4f7b37b82baf2a93e3)

**Comment by AlpinYukseloglu** [comment]
Updates LGTM! Rounding at precision in intermediate calcs might be overkill here (in the original implementation as well) but as long as rounding error doesn't blow up anywhere later I think it's good to keep

**Comment by p0mvn** [comment]
Looks like this resolved the sub-ULP issue we were having pre-launch judging by the failing test that was reproducing it: https://github.com/osmosis-labs/osmosis/blob/f8fb807e9269fad20d611a4f7b37b82baf2a93e3/x/concentrated-liquidity/swapstrategy/one_for_zero_test.go#L186

Going to look into it now
